File: go/subtle/subtleutil/subtleutil_test.go
// Copyright 2017 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////
package subtleutil_test

import (
  "encoding/hex"
  "testing"
  "github.com/google/tink/go/subtle/subtleutil"
)

func TestConvertHashName(t *testing.T) {
  if subtleutil.ConvertHashName("SHA-256") != "SHA256" ||
    subtleutil.ConvertHashName("SHA-1") != "SHA1" ||
    subtleutil.ConvertHashName("SHA-512") != "SHA512" ||
    subtleutil.ConvertHashName("UNKNOWN_HASH") != "" {
    t.Errorf("incorrect hash name conversion")
  }
}

func TestConvertCurveName(t *testing.T) {
  if subtleutil.ConvertCurveName("secp256r1") != "NIST_P256" ||
    subtleutil.ConvertCurveName("secp384r1") != "NIST_P384" ||
    subtleutil.ConvertCurveName("secp521r1") != "NIST_P521" ||
    subtleutil.ConvertCurveName("UNKNOWN_CURVE") != "" {
    t.Errorf("incorrect curve name conversion")
  }
}


func TestComputeHash(t *testing.T) {
  data := []byte("Hello")
  // SHA1
  expectedMac := "f7ff9e8b7bb2e09b70935a5d785e0cc5d9d0abf0"
  hashFunc := subtleutil.GetHashFunc("SHA1")
  if hashFunc == nil || hex.EncodeToString(subtleutil.ComputeHash(hashFunc, data)) != expectedMac {
    t.Errorf("invalid hash function for SHA1")
  }
  // SHA256
  expectedMac = "185f8db32271fe25f561a6fc938b2e264306ec304eda518007d1764826381969"
  hashFunc = subtleutil.GetHashFunc("SHA256")
  if hashFunc == nil || hex.EncodeToString(subtleutil.ComputeHash(hashFunc, data)) != expectedMac {
    t.Errorf("invalid hash function for SHA256")
  }
  // SHA512
  expectedMac = "3615f80c9d293ed7402687f94b22d58e529b8cc7916f8fac7fddf7fbd5af4cf" +
                "777d3d795a7a00a16bf7e7f3fb9561ee9baae480da9fe7a18769e71886b03f315"
  hashFunc = subtleutil.GetHashFunc("SHA512")
  if hashFunc == nil || hex.EncodeToString(subtleutil.ComputeHash(hashFunc, data)) != expectedMac {
    t.Errorf("invalid hash function for SHA512")
  }
  // unknown
  if subtleutil.GetHashFunc("UNKNOWN_HASH") != nil {
    t.Errorf("unexpected result for invalid hash types")
  }
}

func TestGetCurve(t *testing.T) {
  if subtleutil.GetCurve("NIST_P224").Params().Name != "P-224" {
    t.Errorf("incorrect result for NIST_P224")
  }
  if subtleutil.GetCurve("NIST_P256").Params().Name != "P-256" {
    t.Errorf("incorrect result for NIST_P256")
  }
  if subtleutil.GetCurve("NIST_P384").Params().Name != "P-384" {
    t.Errorf("incorrect result for NIST_P384")
  }
  if subtleutil.GetCurve("NIST_P521").Params().Name != "P-521" {
    t.Errorf("incorrect result for NIST_P521")
  }
  if subtleutil.GetCurve("UNKNOWN_CURVE") != nil {
    t.Errorf("expect nil when curve is unknown")
  }
}

File: java_src/src/test/java/com/google/crypto/tink/subtle/SubtleUtilTest.java
// Copyright 2018 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

package com.google.crypto.tink.subtle;

import static com.google.common.truth.Truth.assertThat;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertThrows;

import com.google.crypto.tink.subtle.Enums.HashType;
import java.math.BigInteger;
import java.nio.ByteBuffer;
import java.security.GeneralSecurityException;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

/** Unit tests for SubtleUtil. */
@RunWith(JUnit4.class)
public final class SubtleUtilTest {
  @Test
  public void testToEcdsaAlgo() throws Exception {
    assertEquals("SHA256withECDSA", SubtleUtil.toEcdsaAlgo(HashType.SHA256));
    assertEquals("SHA512withECDSA", SubtleUtil.toEcdsaAlgo(HashType.SHA512));
    assertThrows(GeneralSecurityException.class, () -> SubtleUtil.toEcdsaAlgo(HashType.SHA1));
  }

  @Test
  public void testToRsaSsaPkcs1Algo() throws Exception {
    assertEquals("SHA256withRSA", SubtleUtil.toRsaSsaPkcs1Algo(HashType.SHA256));
    assertEquals("SHA512withRSA", SubtleUtil.toRsaSsaPkcs1Algo(HashType.SHA512));
    assertThrows(GeneralSecurityException.class, () -> SubtleUtil.toRsaSsaPkcs1Algo(HashType.SHA1));
  }

  @Test
  public void testPutAsUnsigedInt_smallNumber() throws Exception {
    ByteBuffer buffer = ByteBuffer.allocate(4);
    SubtleUtil.putAsUnsigedInt(buffer, 0x1122EEFFL);
    assertThat(buffer.array()).isEqualTo(new byte[] {0x11, 0x22, (byte) 0xEE, (byte) 0xFF});
  }

  @Test
  public void testPutAsUnsigedInt_largeNumber() throws Exception {
    ByteBuffer buffer = ByteBuffer.allocate(4);
    SubtleUtil.putAsUnsigedInt(buffer, 0xFFEEDDCCL);
    assertThat(buffer.array())
        .isEqualTo(new byte[] {(byte) 0xFF, (byte) 0xEE, (byte) 0xDD, (byte) 0xCC});
  }

  @Test
  public void testPutAsUnsigedInt_max() throws Exception {
    ByteBuffer buffer = ByteBuffer.allocate(4);
    SubtleUtil.putAsUnsigedInt(buffer, 0xFFFFFFFFL);
    assertThat(buffer.array())
        .isEqualTo(new byte[] {(byte) 0xFF, (byte) 0xFF, (byte) 0xFF, (byte) 0xFF});
  }

  @Test
  public void testPutAsUnsigedInt_tooLargeNumber_throws() throws Exception {
    ByteBuffer buffer = ByteBuffer.allocate(4);
    assertThrows(
        GeneralSecurityException.class, () -> SubtleUtil.putAsUnsigedInt(buffer, 0xFFFFFFFFL + 1L));
  }

  @Test
  public void testPutAsUnsigedInt_minusOne_throws() throws Exception {
    ByteBuffer buffer = ByteBuffer.allocate(4);
    assertThrows(GeneralSecurityException.class, () -> SubtleUtil.putAsUnsigedInt(buffer, -1));
  }

  @Test
  public void bytes2Integer() throws Exception {
    assertThat(SubtleUtil.bytes2Integer(new byte[] {(byte) 0x00}))
        .isEqualTo(BigInteger.ZERO);
    assertThat(SubtleUtil.bytes2Integer(new byte[] {(byte) 0x01}))
        .isEqualTo(BigInteger.ONE);
    assertThat(SubtleUtil.bytes2Integer(new byte[] {(byte) 0x7F}))
        .isEqualTo(BigInteger.valueOf(127));
    // The input should be interpreted as an unsigned integers. So 0x80 is 128.
    assertThat(SubtleUtil.bytes2Integer(new byte[] {(byte) 0x80}))
        .isEqualTo(BigInteger.valueOf(128));
    assertThat(SubtleUtil.bytes2Integer(new byte[] {(byte) 0xFF}))
        .isEqualTo(BigInteger.valueOf(255));
    assertThat(SubtleUtil.bytes2Integer(new byte[] {(byte) 0x01, (byte) 0x00}))
        .isEqualTo(BigInteger.valueOf(256));
    assertThat(SubtleUtil.bytes2Integer(new byte[] {(byte) 0x01, (byte) 0x02}))
        .isEqualTo(BigInteger.valueOf(258));
    // leading zeros are ignored
    assertThat(SubtleUtil.bytes2Integer(
                   new byte[] {(byte) 0x00, (byte) 0x00, (byte) 0x00, (byte) 0x01, (byte) 0x02}))
        .isEqualTo(BigInteger.valueOf(258));
    // the empty array is decoded to 0.
    assertThat(
        SubtleUtil.bytes2Integer(new byte[] {})).isEqualTo(BigInteger.ZERO);
  }

  @Test
  public void integer2Bytes_success() throws Exception {
    assertThat(SubtleUtil.integer2Bytes(BigInteger.ZERO, /*intendedLength=*/ 0))
        .isEqualTo(new byte[] {});
    assertThat(SubtleUtil.integer2Bytes(BigInteger.ZERO, /*intendedLength=*/ 1))
        .isEqualTo(new byte[] {(byte) 0x00});
    assertThat(SubtleUtil.integer2Bytes(BigInteger.ZERO, /*intendedLength=*/ 2))
        .isEqualTo(new byte[] {(byte) 0x00, (byte) 0x00});
    assertThat(SubtleUtil.integer2Bytes(BigInteger.ONE, /*intendedLength=*/ 1))
        .isEqualTo(new byte[] {(byte) 0x01});
    assertThat(SubtleUtil.integer2Bytes(BigInteger.ONE, /*intendedLength=*/ 2))
        .isEqualTo(new byte[] {(byte) 0x00, (byte) 0x01});
    assertThat(SubtleUtil.integer2Bytes(BigInteger.valueOf(127), /*intendedLength=*/ 1))
        .isEqualTo(new byte[] {(byte) 0x7F});
    assertThat(SubtleUtil.integer2Bytes(BigInteger.valueOf(127), /*intendedLength=*/ 2))
        .isEqualTo(new byte[] {(byte) 0x00, (byte) 0x7F});
    assertThat(SubtleUtil.integer2Bytes(BigInteger.valueOf(127), /*intendedLength=*/ 3))
        .isEqualTo(new byte[] {(byte) 0x00, (byte) 0x00, (byte) 0x7F});
    assertThat(SubtleUtil.integer2Bytes(BigInteger.valueOf(128), /*intendedLength=*/ 1))
        .isEqualTo(new byte[] {(byte) 0x80});
    assertThat(SubtleUtil.integer2Bytes(BigInteger.valueOf(128), /*intendedLength=*/ 2))
        .isEqualTo(new byte[] {(byte) 0x00, (byte) 0x80});
    assertThat(SubtleUtil.integer2Bytes(BigInteger.valueOf(128), /*intendedLength=*/ 3))
        .isEqualTo(new byte[] {(byte) 0x00, (byte) 0x00, (byte) 0x80});
    assertThat(SubtleUtil.integer2Bytes(BigInteger.valueOf(255), /*intendedLength=*/ 1))
        .isEqualTo(new byte[] {(byte) 0xFF});
    assertThat(SubtleUtil.integer2Bytes(BigInteger.valueOf(255), /*intendedLength=*/ 2))
        .isEqualTo(new byte[] {(byte) 0x00, (byte) 0xFF});
    assertThat(SubtleUtil.integer2Bytes(BigInteger.valueOf(255), /*intendedLength=*/ 3))
        .isEqualTo(new byte[] {(byte) 0x00, (byte) 0x00, (byte) 0xFF});
    assertThat(SubtleUtil.integer2Bytes(BigInteger.valueOf(256), /*intendedLength=*/ 2))
        .isEqualTo(new byte[] {(byte) 0x01, (byte) 0x00});
    assertThat(SubtleUtil.integer2Bytes(BigInteger.valueOf(258), /*intendedLength=*/ 2))
        .isEqualTo(new byte[] {(byte) 0x01, (byte) 0x02});
    assertThat(SubtleUtil.integer2Bytes(BigInteger.valueOf(258), /*intendedLength=*/ 4))
        .isEqualTo(new byte[] {(byte) 0x00, (byte) 0x00, (byte) 0x01, (byte) 0x02});
  }

  @Test
  public void integer2Bytes_failWhenIntegerIsNegative() throws Exception {
    assertThrows(
        IllegalArgumentException.class,
        () -> SubtleUtil.integer2Bytes(BigInteger.valueOf(-1), /*intendedLength=*/ 2));
    assertThrows(
        IllegalArgumentException.class,
        () -> SubtleUtil.integer2Bytes(BigInteger.valueOf(-42), /*intendedLength=*/ 2));
  }

  @Test
  public void integer2Bytes_failWhenIntegerIsLargerThanIntendedLength() throws Exception {
    assertThrows(
        GeneralSecurityException.class,
        () -> SubtleUtil.integer2Bytes(BigInteger.ONE, /* intendedLength= */ 0));
    assertThrows(
        GeneralSecurityException.class,
        () -> SubtleUtil.integer2Bytes(BigInteger.valueOf(256), /*intendedLength=*/ 1));
    assertThrows(
        GeneralSecurityException.class,
        () -> SubtleUtil.integer2Bytes(BigInteger.valueOf(256 * 256), /*intendedLength=*/ 2));
  }

}
